tfe/resource_tfe_policy_set.go

```diff
@@ -0,0 +1,277 @@
+	global := d.Get("global").(bool)
+
+	// If a user is setting the policy set to "global", make sure the workspaces
+	// that _had_ been set are explicitly removed. This helps keep the policy
+	// set's state in check
+	if d.HasChange("global") && global {
+		// The new set of workspaces will be an empty set, so we don't need it
+		oldSet, _ := d.GetChange("workspace_external_ids")
+		oldWorkspaceIDs := oldSet.(*schema.Set)
+
+		if oldWorkspaceIDs.Len() > 0 {
+			options := tfe.PolicySetDetachFromWorkspacesOptions{}
+
+			for _, workspaceID := range oldWorkspaceIDs.List() {
+				options.Workspaces = append(options.Workspaces, &tfe.Workspace{ID: workspaceID.(string)})
+			}
+
+			log.Printf("[DEBUG] Removing previous workspaces from now-global policy set: %s", d.Id())
+			err := tfeClient.PolicySets.DetachFromWorkspaces(ctx, d.Id(), options)
+			if err != nil {
+				return fmt.Errorf("Error detaching policy set %s from workspaces: %v", d.Id(), err)
+			}
+		}
+	}
+
+	// Don't bother updating the policy set's attributes if they haven't changed
+	if d.HasChange("name") || d.HasChange("description") || d.HasChange("global") {
+		// Create a new options struct.
+		options := tfe.PolicySetUpdateOptions{
+			Name:   tfe.String(name),
+			Global: tfe.Bool(global),
+		}
+
+		if desc, ok := d.GetOk("description"); ok {
+			options.Description = tfe.String(desc.(string))
+		}
+
+		log.Printf("[DEBUG] Update configuration for policy set: %s", d.Id())
+		_, err := tfeClient.PolicySets.Update(ctx, d.Id(), options)
+		if err != nil {
+			return fmt.Errorf(
+				"Error updating configuration for policy set %s: %v", d.Id(), err)
+		}
+	}
+
+	if d.HasChange("policy_ids") {
+		oldSet, newSet := d.GetChange("policy_ids")
+		oldPolicyIDs := oldSet.(*schema.Set).Difference(newSet.(*schema.Set))
+		newPolicyIDs := newSet.(*schema.Set).Difference(oldSet.(*schema.Set))
+
+		// First add the new policies.
+		if newPolicyIDs.Len() > 0 {
+			options := tfe.PolicySetAddPoliciesOptions{}
+
+			for _, policyID := range newPolicyIDs.List() {
+				options.Policies = append(options.Policies, &tfe.Policy{ID: policyID.(string)})
+			}
+
+			log.Printf("[DEBUG] Add policies to policy set: %s", d.Id())
+			err := tfeClient.PolicySets.AddPolicies(ctx, d.Id(), options)
+			if err != nil {
+				return fmt.Errorf("Error adding policies to policy set %s: %v", d.Id(), err)
+			}
+		}
+
+		// Then remove all the old policies.
+		if oldPolicyIDs.Len() > 0 {
+			options := tfe.PolicySetRemovePoliciesOptions{}
+
+			for _, policyID := range oldPolicyIDs.List() {
+				options.Policies = append(options.Policies, &tfe.Policy{ID: policyID.(string)})
+			}
+
+			log.Printf("[DEBUG] Remove policies from policy set: %s", d.Id())
+			err := tfeClient.PolicySets.RemovePolicies(ctx, d.Id(), options)
+			if err != nil {
+				return fmt.Errorf("Error removing policies from policy set %s: %v", d.Id(), err)
+			}
+		}
+	}
+
+	if !global && d.HasChange("workspace_external_ids") {
```

Comment:
I know you had a question about this block and the above one, specifically with regards to why both `if` check blocks are needed… I'll try to walk you through why I've separated these two pieces out like this, and why the first check is also the first thing we do in this function. It largely has to do with the way the API works, as well as trying to avoid a specific scenario. Consider the following steps:

1. I have an organization that has five workspaces. `ws-123`, `ws-456`, `ws-789`, `ws-abc` and `ws-def`
2. I create a policy set with the default `global = false` and specify a couple of these workspaces as set members (`ws-123` and `ws-789`).
3. Later, I revisit this and decide to make the policy set global. I set `global = true` and remove the `workspace_external_ids` attribute entirely.

Without the top statement, we would:

1. Update the policy set itself (name/description/global flag)
2. Update the list of policies
3. And then we'd get to this block I've highlighted here and attempt to "update" the workspaces (in this case, that's just changed to an empty set so we would try to submit an API request that removes `ws-123` and `ws-789` from the set)

However, once a set has been flagged as global, the TFE API does not allow directly modifying the set's workspaces relationship; this API request would error out. So we have to handle the removal before we submit the request that actually flags the set as global.

I also can't move this block to the top in favor of the one already there, because it messes with what is essentially the reverse of the scenario I outlined at the start of this comment (i.e. changing from being global). For example: I remove the `global = true` statement from my policy set and then specify a few workspaces in the `workspace_external_ids` property. If we tried to just handle this in one combined check at the top, we would:

1. Detect the change to `workspace_external_ids` and attempt to submit an API request that adds the workspace
2. Get an error back from the API because we have not yet submitted the API request that unflags the set as global.

So we have to use a different order of operations depending on whether a set is changing global from `false => true` or from `true => false`.

I hope this long-winded comment made sense and adequately explained why everything is ordered this way 😅

Reply:
Thanks for the long and detailed explanation! I does raise another question, but I'll first merge this in and then have a look at the other thought this triggers.

Thanks for your work and patience!!